### PR TITLE
Fix FAQ/KB

### PR DESCRIPTION
**Pull request by AdrienClairembault**
The FAQ / KB view does not seem to work properly.
You can see on the following images that the category tree indicate that some articles should be visible, however when you click on the category you find out that there is no valid articles:

image
image
image

The code handling this list was hard to read because it was handling two different cases (FAQ and KB) at the same time.

I've: 
* Split and moved the code around to have proper KB / FAQ "code paths" which are easier to read and understand.
* Added a great number of tests which should handle every possible KB configuration.

Regarding the specific error in the images above, I think the `is_faq => 1` condition was not properly applied for self-service users (it was thrown in an `OR` clause with other criteria that should only be used for the KB side, not the FAQ).
The new code handle it correctly as it separate KB and FAQ more cleanly, thus only applying the correct condition.

More details on the FAQ / KB expected restrictions can be found here : https://glpi-user-documentation.readthedocs.io/fr/latest/modules/tools/knowledgebase.html
> Only public FAQ items are visible to users of simplified interface. 
FAQ items are public to everyone (within their entity) that has access to read FAQs.
The additional visibility restrictions will have no affect in this case

> Other elements are visible only to technicians via standard interface

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !25026

### CODE REVIEW

File: tests/functionnal/KnowbaseItem.php [hunk 1]
```diff
@@ -508,4 +509,323 @@ public function testCreateWithCategories()
...
+
+        // First FAQ test case: public FAQ disabled
+        $CFG_GLPI['use_public_faq'] = false;
+        yield ['articles' => []];
+
+        // Second FAQ test case: public FAQ enabled + multi entities
+        if (!isset($_SESSION['glpi_multientitiesmode'])) {
+            $_SESSION['glpi_multientitiesmode'] = 1;
+        }
+        $this->integer($_SESSION['glpi_multientitiesmode'])->isEqualTo(1);
```
Comment (cedric-anne) [inline tests/functionnal/KnowbaseItem.php:566]:
```suggestion
        $_SESSION['glpi_multientitiesmode'] = 1;
```

File: tests/functionnal/KnowbaseItem.php [hunk 2]
```diff
@@ -508,4 +509,323 @@ public function testCreateWithCategories()
...
+        $kb_4 = getItemByTypeName("KnowbaseItem", "KB 4", true);
+        $kb_5 = getItemByTypeName("KnowbaseItem", "KB 5", true);
+        $kb_6 = getItemByTypeName("KnowbaseItem", "KB 6", true);
+        $kb_7 = getItemByTypeName("KnowbaseItem", "KB 7", true);
+        $group_a = $this->createItem("Group", [
+            "name" => "Group KB A",
+            'is_recursive' => 1,
+        ])->fields['id'];
+        $group_b = $this->createItem("Group", [
+            "name" => "Group KB A",
```
Comment (cedric-anne) [inline tests/functionnal/KnowbaseItem.php:695]:
```suggestion
            "name" => "Group KB B",
```

File: tests/functionnal/KnowbaseItem.php [hunk 3]
```diff
@@ -508,4 +509,323 @@ public function testCreateWithCategories()
         }
         $this->array($category_ids)->containsValues([$kb_cat_id1, $kb_cat_id2]);
     }
+
+    protected function testGetVisibilityCriteriaProvider(): Generator
```
[the remark this answers is not in the document]
Comment (AdrienClairembault) [reply tests/functionnal/KnowbaseItem.php:513]: A generator is an iterable but an iterable isn't always a Generator.
Doesn't it make sense to use the more precise return type ?

File: src/KnowbaseItem.php [hunk 4]
```diff
@@ -572,22 +572,66 @@ public static function getVisibilityCriteria(bool $forceall = false): array
...
+        if (!Session::getLoginUserID()) {
+            if ($CFG_GLPI["use_public_faq"]) {
+                // Public FAQ is enabled; show FAQ
+                $criteria['WHERE'] = self::getVisibilityCriteriaFAQ();
+                return $criteria;
+            } else {
+                // Public FAQ is disabled; show nothing
+                $criteria['WHERE'] = [0];
+                return $criteria;
+            }
```
Comment (btry) [inline src/KnowbaseItem.php:590]: for readability : limiting nest level 

```suggestion
            }
             // Public FAQ is disabled; show nothing
             $criteria['WHERE'] = [0];
             return $criteria;
```

File: src/KnowbaseItem.php [hunk 5]
```diff
@@ -572,22 +572,66 @@ public static function getVisibilityCriteria(bool $forceall = false): array
...
+        // Handle logged in users
+        if (Session::getCurrentInterface() == "helpdesk") {
+            // Show FAQ for helpdesk user
+            $criteria['WHERE'] = self::getVisibilityCriteriaFAQ();
+            return $criteria;
+        } else {
+            // Show knowledge base for central users
+            $criteria['WHERE'] = self::getVisibilityCriteriaKB();
+            return $criteria;
+        }
```
Comment (btry) [inline src/KnowbaseItem.php:602]: Same idea here : reduce code nesting 
```suggestion
        }
         // Show knowledge base for central users
         $criteria['WHERE'] = self::getVisibilityCriteriaKB();
         return $criteria;
```

File: src/KnowbaseItem.php (hunk 4 above)
Comment (AdrienClairembault) [reply src/KnowbaseItem.php:590]: If/else is more readable to me, I don't think this is important.

File: tests/functionnal/KnowbaseItem.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (cedric-anne) [reply tests/functionnal/KnowbaseItem.php:513]: 1. This is a test data provider method. Having to declare a use statement just for this seems too much for me.
2. `atoum` requires test data provider methods to return any `iterable` type, so there is no need to restrict it to a `Generator` for anyone that would want to rewrite the content of this method in the future.

Anyway, this is a test data provider method, so type hinting on its return type is not a big deal.

**Review by cedric-anne — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Comment by AdrienClairembault** [comment]
Wait before merge, I've got one final check to do following internal ref feedback

**Review by Rom1-B — APPROVED** [review]

**Review by btry — APPROVED** [review]

**Pull request merged**